src/AWS.Logger.Core/Core/AWSLoggerCore.cs

```diff
@@ -103,48 +117,112 @@ private static AWSCredentials DetermineCredentials(AWSLoggerConfig config)
 
         public void Close()
         {
-            _isTerminated = true;
-            _cancelStartSource.Cancel();
-            _config.ShutDown();
-            Task.Run(async () =>
+            try
+            {
+                Flush();
+                _cancelStartSource.Cancel();
+            }
+            catch (Exception ex)
             {
-                await Monitor(CancellationToken.None);
-            }).Wait();
+                LogLibraryServiceError(ex);
+            }
+            finally
+            {
+                LogLibraryAlert = null;
+            }
         }
 
-
-        /// <summary>
-        /// A Concurrent Queue is used to store the messages from 
-        /// the logger
-        /// </summary>
-        /// <param name="message"></param>
-        public void AddMessage(string rawMessage)
+        public void Flush()
         {
-            Action<string> processor = (message) =>
+            if (_cancelStartSource.IsCancellationRequested)
+                return;
+
+            if (!_pendingMessageQueue.IsEmpty || !_repo.IsEmpty)
             {
-                if (_pendingMessageQueue.Count > _config.MaxQueuedMessages)
+                bool lockTaken = false;
+                try
                 {
-                    if ((_maxBufferTimeStamp == DateTime.MinValue) || (DateTime.Now > _maxBufferTimeStamp.Add(TimeSpan.FromMinutes(MAX_BUFFER_TIMEDIFF))))
+                    // Ensure only one thread executes the flush operation
+                    System.Threading.Monitor.TryEnter(_flushTriggerEvent, ref lockTaken);
+                    if (lockTaken)
                     {
-                        _maxBufferTimeStamp = DateTime.Now;
-                        message = "The AWS Logger in-memory buffer has reached maximum capacity";
-                        _pendingMessageQueue.Enqueue(new InputLogEvent
+                        _flushCompletedEvent.Reset();
+                        if (_flushTriggerEvent.CurrentCount == 0)
```

Comment:
As soon the background worker task has performed credential handshake and waits on the SemaphoreSlim then it becomes 1. When 0 it means that the background-worker is not yet ready for flush.

Reply:
In a very short lived process like the [AspnetCore/ConsoleSample](https://github.com/aws/aws-logging-dotnet/tree/master/samples/AspNetCore/ConsoleSample) the flush is called before the initial send is complete so I'm seeing it go into the error case. In what case would we not want the background task to flush the messages if the flush method has been called. I'm not sure I understand what it would be busy with. In the case of the ConsoleSample we would want it to flush as soon as the LogStream was set up.

Reply:
@normj I have now fixed the bug, see my last commit.